sycl/doc/extensions/proposed/sycl_ext_oneapi_record_event.asciidoc

```diff
@@ -0,0 +1,219 @@
+get timing information for some commands, but the library does not control the
+construction of the queue (which is where the `enable_profiling` property is
+passed).
+
+This extension is structured as a free function, rather than a member function
+on `queue`, in order to be consistent with the API design in
+link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
+sycl_ext_oneapi_enqueue_functions]
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_RECORD_EVENT` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New device aspect
+
+This extension adds the `ext_oneapi_queue_event_recording` enumerator to the
+`sycl::aspect` enumeration.
+
+```
+namespace sycl {
+
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_queue_event_recording
+};
+
+} // namespace sycl
+```
+
+When a device has this aspect, the `record_event` function may be called for a
+queue on this device even if the queue is not constructed with the property
+`property::queue::enable_profiling`.
+
+=== New free function
+
+This extension adds the following free function.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+event record_event(const queue& q);
+
+} // namespace sycl::ext::oneapi::experimental
+----
+!====
+
+_Effects:_ Enqueues a command barrier to `q`.
+Any commands submitted after this barrier cannot begin execution until all
+previously submitted commands have completed.
+
+_Returns:_ An event which represents the completion of the barrier.
+The event's status becomes `info::event_command_status::complete` when all
+commands submitted to the queue prior to the call to `record_event` have
+completed.
+The event's `info::event_profiling::command_submit` timestamp reflects the
+time at which `record_event` is called.
+The event's `info::event_profiling::command_end` timestamp reflects the time
+at which the event enters the "complete" state.
+
+It is unspecified whether the event ever has the
+`info::event_command_status::running` status, and the meaning of the
+`info::event_profiling::command_start` timestamp is also unspecified.
+Implementations are encouraged to transition the event directly from the
+"submitted" status to the "complete" status and are encouraged to set the
+"command_start" timestamp to the same value as the "command_end" timestamp.
```

Comment:
I wonder if we should require this instead of encouraging it, for consistency and portability of generic profiling code.

Reply:
The previous wording did require "start" and "end" timestamps to be the same, but @Pennycook convinced me to weaken the language in order to accommodate backends that don't have native barrier support.  I think generic profiling code is still portable.  Note how the example shows to calculate the time between two events.  The first timestamp is the "end" of the first event, and the second timestamp is the "start" of the second event.  Therefore, it doesn't matter if "start" and "end" are the same, any difference is excluded from the computed elapsed time.